File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        bracketPair = {")":"(","}":"{","]":"["}

        for c in s:
            if c in bracketPair:
                if stack and stack[-1]== bracketPair[c]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)
        
        return len(stack) == 0

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        //use stack and a map
        map<char,char> m;
        m.insert({'[',']'});
        m.insert({'(',')'});
        m.insert({'{','}'});

        stack<char>st;

        for(auto &c: s){
            if(m.find(c)!=m.end()){
                st.push(c);
            }
            else {
                if(st.empty()||m[st.top()]!=c){
                    return false;
                }      
                st.pop();
            }
        }
        return st.empty();
    }
};